### PR TITLE
Use TemplateMatcher for ODataBatchPathMapping

### PR DESCRIPTION
**Pull request by xuzhg**
### Issues

*This pull request fixes issue https://github.com/OData/WebApi/issues/1386.* It's a feature request.

### Description

* Use TemplateMatcher for ODataBatchPathMapping
* Separate the HttpRequestMessageExtensions into different project (classic and core)

### Checklist (Uncheck if it is not completed)

- [x] *Test cases added*
- [x] *Build and test with one-click build and test script passed*

### Additional work necessary

*If documentation update is needed, please add "Docs Needed" label to the issue and provide details about the required document change in the issue.*

### CODE REVIEW

File: test/UnitTest/Microsoft.AspNetCore.OData.Test/Extensions/HttpRequestExtensions.cs [hunk 1]
```diff
@@ -1,22 +1,13 @@
...
 using Microsoft.AspNetCore.Http;
-#else
-using System.Net.Http;
-using Microsoft.AspNet.OData.Extensions;
-#endif
 
 namespace Microsoft.AspNet.OData.Test.Extensions
 {
-#if NETCORE
-    /// <summary>
```
Comment (biaol-odata) [inline test/UnitTest/Microsoft.AspNetCore.OData.Test/Extensions/HttpRequestExtensions.cs]: good to keep for AspNetCore?

Comment (xuzhg) [reply test/UnitTest/Microsoft.AspNetCore.OData.Test/Extensions/HttpRequestExtensions.cs]: This file is separated into two files, one in the classic, one in the AspNetCore.

File: src/Microsoft.AspNetCore.OData/Interfaces/IODataFeature.cs [hunk 2]
```diff
@@ -42,6 +43,11 @@ public interface IODataFeature
         /// </summary>
         Uri NextLink { get; set; }
 
+        /// <summary>
+        /// Gets or sets the batch rote data.
```
[the remark this answers is not in the document]
Comment (xuzhg) [reply src/Microsoft.AspNetCore.OData/Interfaces/IODataFeature.cs]: Thanks.

File: test/UnitTest/Microsoft.AspNetCore.OData.Test/Extensions/HttpRequestExtensions.cs (hunk 1 above)
Comment (xuzhg) [reply test/UnitTest/Microsoft.AspNetCore.OData.Test/Extensions/HttpRequestExtensions.cs]: Ok. Got it. Thanks.

**Pull request merged**